src/workerd/api/actor-state.c++

```diff
@@ -713,6 +715,61 @@ jsg::Promise<jsg::Value> DurableObjectState::blockConcurrencyWhile(jsg::Lock& js
   return IoContext::current().blockConcurrencyWhile(js, kj::mv(callback));
 }
 
+void DurableObjectState::acceptWebSocket(
+    jsg::Ref<WebSocket> ws,
+    jsg::Optional<kj::Array<kj::String>> tags) {
+  JSG_ASSERT(!ws->isAccepted(), Error,
+      "Cannot call `acceptWebSocket()` if the WebSocket was already accepted via `accept()`");
+  JSG_ASSERT(ws->LOCAL == WebSocket::Locality::LOCAL, Error,
+      "Outgoing WebSocket connections cannot be hibernated.");
+
+  // We need to get a HibernationManager to give the websocket to.
+  auto& a = KJ_REQUIRE_NONNULL(IoContext::current().getActor());
+  if (a.getHibernationManager() == nullptr) {
+    a.setHibernationManager(
+        kj::refcounted<HibernationManagerImpl>(
+            a.getLoopback(), KJ_REQUIRE_NONNULL(a.getHibernationEventType())));
+  }
+  // HibernationManager's acceptWebSocket() will throw if the websocket is in an incompatible state.
+  // Note that not providing a tag is equivalent to providing an empty tag array.
+  // Any duplicate tags will be ignored.
+  kj::Array<kj::String> distinctTags = [&]() -> kj::Array<kj::String> {
+    KJ_IF_MAYBE(t, tags) {
+      kj::HashSet<kj::String> seen;
+      for (auto tag = t->begin(); tag < t->end(); tag++) {
+        // TODO(now): Need to make sure this error makes its way up to the worker script.
+        KJ_REQUIRE(tag->size() <= 256, "the maximum tag length is 256 characters!");
+        if (!seen.contains(*tag)) {
+          seen.insert(kj::mv(*tag));
+        }
+      }
+
+      // TODO(now): Need to make sure this error makes its way up to the worker script.
+      KJ_REQUIRE(seen.size() <= 10, "a Hibernatable WebSocket cannot have more than 10 tags");
```

Comment:
Couldn't this be a counter like `added` and when > 10 throw the error, instead of reading all tags and only then do the `KJ_REQUIRE`?

Reply:
Yes, good idea!